mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -387,29 +386,30 @@ private static void addSubQuery(String operator, BasicDBObject query, String col
 		query.append( operator, subQuery.append( column, value ) );
 	}
 
-	@Override
-	public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext) {
-		BasicDBObject idObject = prepareIdObject( key );
+	private static void addSetToQuery(BasicDBObject query, String column, Object value) {
+		removeSubQuery( "$unset", query, column );
+		addSubQuery( "$set", query, column, value );
+	}
 
-		DBObject updater = objectForUpdate( tuple, idObject, tupleContext );
-		WriteConcern writeConcern = getWriteConcern( tupleContext );
+	private static void addUnsetToQuery(BasicDBObject query, String column) {
+		removeSubQuery( "$set", query, column );
+		addSubQuery( "$unset", query, column, Integer.valueOf( 1 ) );
+	}
 
-		try {
-			getCollection( key ).update( idObject, updater, true, false, writeConcern );
-		}
-		catch ( DuplicateKeyException dke ) {
-			// This exception is used by MongoDB for all the unique indexes violation, not only the primary key
-			// so we determine if it concerns the primary key by matching on the message
-			if ( PRIMARY_KEY_CONSTRAINT_VIOLATION_MESSAGE.matcher( dke.getMessage() ).matches() ) {
-				throw new TupleAlreadyExistsException( key.getMetadata(), tuple, dke );
-			}
-			else {
-				throw log.constraintViolationForEntity( key, dke.getMessage(), dke );
-			}
+	private static void removeSubQuery(String operator, BasicDBObject query, String column) {
+		BasicDBObject subQuery = getSubQuery( operator, query );
+		subQuery.removeField( column );
+		if ( subQuery.isEmpty() ) {
+			query.removeField( operator );
 		}
 	}
 
 	@Override
+	public void insertOrUpdateTuple(EntityKey key, TuplePointer tuplePointer, TupleContext tupleContext) {
+		throw new UnsupportedOperationException( "Method not supported in GridDialect anymore" );
+	}
```

Comment:
Can we move this in `AbstractGroupingByEntityDialect` as a default implementation?